### PR TITLE
Don't hold dashmap write lock in store create

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
@sakridge  noticed that `get_slot_storages()` in AccountsDb was occasionally very slow: 
Screen Shot 2020-10-19 at 11 22 09 PM

#### Summary of Changes
At first I thought it might be some bug in `DashMap`, but I ran

1) The benchmarks as included in this PR here
2) Variations of this benchmark with multiple readers/no writers
3) Varying numbers of keys in the maps
4) Reading randomly chosen keys in the map instead of the same key 

And throughout DashMap seemed to outperform/perform as well as the `Arc<RwLock>`

Turns out the issue is holding the write lock on the dashmap while waiting on a separate write lock here: https://github.com/solana-labs/solana/blob/942e4273ba52cade3d0a3e93dd7661230c98466f/runtime/src/accounts_db.rs#L1418-L1421, where the second write lock is taking a while (probably contention with reads?). This makes me think it might be worthwhile to turn the inner SlotStore into a `DashMap` as well 😄 

The simplest fix is just add a clone here: https://github.com/solana-labs/solana/blob/942e4273ba52cade3d0a3e93dd7661230c98466f/runtime/src/accounts_db.rs#L1413-L1417 (tested and the results look good), but I think you can also avoid taking the write lock every time as this PR implements.

After results:
Screen Shot 2020-10-19 at 11 34 44 PM

Fixes #

### CODE REVIEW

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -130,9 +130,8 @@ impl AccountStorage {
         slot: Slot,
         store_id: AppendVecId,
     ) -> Option<Arc<AccountStorageEntry>> {
-        self.0
-            .get(&slot)
-            .and_then(|storage_map| storage_map.value().read().unwrap().get(&store_id).cloned())
+        self.get_slot_stores(slot)
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:133]: maybe is this proactive measure for the unrelated possible slow down?

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -1410,12 +1409,18 @@ impl AccountsDB {
         let store =
             Arc::new(self.new_storage_entry(slot, &Path::new(&self.paths[path_index]), size));
         let store_for_index = store.clone();
-        let slot_storage = self
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1413]: DashMap returns RefMut here and the `.clone()` in new code makes it (and the write lock on DashMap backing it) drop as soon as it goes out of scope (well temporary expression in this case), right?

Could you comment about that?

Well, so hard finding and we need some commenting I guess.

Comment (ryoqun) [reply runtime/src/accounts_db.rs:1413]: Anyway, thanks for finding these very naughty bugs!

**Comment by sakridge** [comment]
A bit unfortunate that `DashMap` has these performance pitfalls, but good find.

File: runtime/src/accounts_db.rs (hunk 2 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:1413]: Added a comment and a type for clarity!

File: runtime/src/accounts_db.rs (hunk 1 above)
Comment (carllin) [reply runtime/src/accounts_db.rs:133]: yeah this was just something else I suspected while investigating the root cause

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -1410,12 +1409,18 @@ impl AccountsDB {
...
-            .storage
-            .0
-            .entry(slot)
-            .or_insert(Arc::new(RwLock::new(HashMap::new())));
-        slot_storage
+        let mut slot_storages = self.storage.get_slot_stores(slot);
```
[the remark this answers is not in the document]
Comment (carllin) [reply runtime/src/accounts_db.rs:1412]: fixed, thanks!

File: runtime/src/accounts_db.rs [hunk 4]
```diff
@@ -1410,12 +1409,23 @@ impl AccountsDB {
...
-            .0
-            .entry(slot)
-            .or_insert(Arc::new(RwLock::new(HashMap::new())));
-        slot_storage
+            .get_slot_stores(slot)
+            .or_else(||
+            // DashMap entry.or_insert() returns a RefMut, essentially a write lock,
+            // which is dropped after this block ends, minimizing time held by the lock.
+            // However, we still want to persist the reference to the `SlotStores` behind
+            // the lock, hence we clone it out, (`SlotStores` is an Arc so is cheap to clone).
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1420]: thanks for explaining this. :)

File: runtime/src/accounts_db.rs [hunk 5]
```diff
@@ -1410,12 +1409,23 @@ impl AccountsDB {
...
+            Some(self.storage
+                .0
+                .entry(slot)
+                .or_insert(Arc::new(RwLock::new(HashMap::new())))
+                .clone()))
+            .unwrap();
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1426]: well, I have a tendency of being wary of plain `unwrap()`. How about using `unwrap_or_else`? Well, I should have say so with my original comment...

File: runtime/benches/accounts.rs [hunk 6]
```diff
@@ -145,54 +150,50 @@ fn bench_delete_dependencies(bencher: &mut Bencher) {
 
 #[bench]
 #[ignore]
-fn bench_concurrent_read_write(bencher: &mut Bencher) {
```
Comment (ryoqun) [inline runtime/benches/accounts.rs:148]: how about reverting back these nice AccountsDb-level bench tests? I think keeping and maintain these doesn't cost us much and might be handy to lay around the source tree in addition to the shinny direct DashMap/RwLock<HashMap> showdown. 

I guess, you just rewrote the body of these bench body code in a haste? ;)

File: runtime/src/accounts_db.rs [hunk 7]
```diff
@@ -1410,12 +1409,23 @@ impl AccountsDB {
         let store =
             Arc::new(self.new_storage_entry(slot, &Path::new(&self.paths[path_index]), size));
         let store_for_index = store.clone();
-        let slot_storage = self
+
+        let slot_storages: SlotStores = self
```
Comment (ryoqun) [inline runtime/src/accounts_db.rs:1413]: `: SlotStore  = `: i like the taste of well nuanced explicit type annotation at the age of rampant type inference xD

File: runtime/src/accounts_db.rs (hunk 5 above)
Comment (carllin) [reply runtime/src/accounts_db.rs]: thanks, good catch!

File: runtime/benches/accounts.rs (hunk 6 above)
Comment (carllin) [reply runtime/benches/accounts.rs]: thanks, good catch!

**Review by ryoqun — APPROVED** [review]
lgtm! super good find!

**Review by sakridge — APPROVED** [review]

**Pull request merged**